aten/src/ATen/core/jit_type.h

```diff
@@ -1123,6 +1126,88 @@ struct CAFFE2_API TupleType : public NamedType {
   std::shared_ptr<FunctionSchema> schema_;
 };
 
+struct EnumType;
+using EnumTypePtr = std::shared_ptr<EnumType>;
+struct CAFFE2_API EnumType : public NamedType {
+  friend struct Type;
+  static const TypeKind Kind = TypeKind::EnumType;
+
+  static EnumTypePtr create(const c10::QualifiedName& qualified_name, TypePtr value) {
+    switch (value->kind()) {
+      case TypeKind::AnyType:
+      case TypeKind::IntType:
+      case TypeKind::FloatType:
+      case TypeKind::StringType:
+      case TypeKind::TensorType:
+        return EnumTypePtr(new EnumType(qualified_name, value));
```

Comment:
Didn't think of this earlier, but I don't think we should support mutable Enum values. I don't think it's a common use case, and we would have to maintain one singleton IValue.
```
class Color(Enum):
    RED = torch.tensor(1)

# all Color.RED instances would have to be updated, or we have to maintain a single instance
Color.RED += 2
```

For the same reason, I don't know if we should support `Any` because it could be a mutable type (and it's also not a common use case).